### PR TITLE
refactor(dap): use rpc's sexp encoding

### PR DESCRIPTION
**Pull request by rgrinberg**
avoids the need to write serialize/deserialize functions manually

also remove the whole serializable thing. the serialization format is
already decided by rpc to be sexp.

Note that this doesn't yet switch the action plugin to RPC, but it's the first step as this is what RPC uses for serializing requests.

### CODE REVIEW

**Review by Alizter — APPROVED** [review]

File: otherlibs/action-plugin/src/protocol.ml [hunk 1]
```diff
@@ -24,16 +14,20 @@ module Dependency = struct
...
-      | Glob { path; glob } -> List [ Atom "glob"; Atom path; Atom glob ]
-
-    let t_of_sexp : Sexp.t -> _ = function
-      | List [ Atom "file"; Atom path ] -> Some (File path)
-      | List [ Atom "directory"; Atom path ] -> Some (Directory path)
-      | List [ Atom "glob"; Atom path; Atom glob ] -> Some (Glob { path; glob })
-      | _ -> None
+    let conv =
+      let open Conv in
+      let file = constr "File" string (fun s -> File s) in
```
Comment (snowleopard) [inline otherlibs/action-plugin/src/protocol.ml:19]: Why did the capitalisation change? Was this intentional?

File: otherlibs/action-plugin/src/protocol.ml [hunk 2]
```diff
@@ -123,22 +113,24 @@ module Response = struct
...
-        List [ Atom "need_more_deps"; Dependency.Set.sexp_of_t deps ]
-
-    let t_of_sexp : Sexp.t -> _ = function
-      | List [ Atom "done" ] -> Some Done
-      | List [ Atom "need_more_deps"; sexp ] ->
-        Option.O.(Dependency.Set.t_of_sexp sexp >>| fun xs -> Need_more_deps xs)
-      | _ -> None
+    let conv =
+      let open Conv in
+      let done_ = constr "Done" unit (fun () -> Done) in
```
Comment (snowleopard) [inline otherlibs/action-plugin/src/protocol.ml:118]: Same question here.

File: otherlibs/action-plugin/src/sexpable_intf.ml [hunk 3]
```diff
@@ -1,9 +1,44 @@
-open Stdune
+open Import
+
+module Deserialization_error = struct
```
Comment (snowleopard) [inline otherlibs/action-plugin/src/sexpable_intf.ml:3]: The word "deserialization" seems weird in this context. Perhaps, `Sexp_conversion_error` would be more suitable?

File: otherlibs/action-plugin/src/sexpable_intf.ml [hunk 4]
```diff
@@ -1,9 +1,44 @@
...
+
+module type S = sig
+  type t
+
+  val conv : t Conv.value
+
+  val version : int
+end
+
+module Make (TypeToSerialize : S) = struct
```
Comment (snowleopard) [inline otherlibs/action-plugin/src/sexpable_intf.ml:25]: And here we could use `TypeToConvert`.

File: otherlibs/action-plugin/src/protocol.ml (hunk 1 above)
Comment (rgrinberg) [reply otherlibs/action-plugin/src/protocol.ml:19]: Yes, it's the convention we use in RPC. Though if you want, I can revert this change.

File: otherlibs/action-plugin/src/sexpable_intf.ml (hunk 3 above)
Comment (rgrinberg) [reply otherlibs/action-plugin/src/sexpable_intf.ml:3]: It includes the `Version_mismatch` constructor though, and that's not really a conversion error from Sexp.

File: otherlibs/action-plugin/src/protocol.ml (hunk 1 above)
Comment (snowleopard) [reply otherlibs/action-plugin/src/protocol.ml:19]: OK, fine if this is Dune RPC's convention. It seems a bit inconsistent with Dune language but it's hardly worth debating over.

File: otherlibs/action-plugin/src/sexpable_intf.ml (hunk 3 above)
Comment (snowleopard) [reply otherlibs/action-plugin/src/sexpable_intf.ml:3]: Maybe just `Error` then? Not really important though.

**Review by snowleopard — APPROVED** [review]
Looks good, thanks!

File: otherlibs/action-plugin/src/sexpable_intf.ml (hunk 3 above)
Comment (rgrinberg) [reply otherlibs/action-plugin/src/sexpable_intf.ml:3]: Yeah, that works.

**Pull request merged**